aws/middleware/user_agent.go

```diff
@@ -0,0 +1,121 @@
+package middleware
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"runtime"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/awslabs/smithy-go/httpbinding"
+	"github.com/awslabs/smithy-go/middleware"
+	smithyhttp "github.com/awslabs/smithy-go/transport/http"
+)
+
+const execEnvVar = `AWS_EXECUTION_ENV`
+const execEnvUAKey = `exec-env`
+
+// UserAgent is an interface for modifying the User-Agent of a request.
+type UserAgent interface {
+	AddKey(name string)
+	AddKeyValue(name, version string)
+}
+
+// RequestUserAgent is a build middleware that set the User-Agent for the request.
+type RequestUserAgent struct {
+	uab *httpbinding.UserAgentBuilder
+}
+
+// NewRequestUserAgent returns a new RequestUserAgent which will set the User-Agent for the request.
+//
+// Default Example:
+//   aws-sdk-go/2.3.4 GOOS/linux GOARCH/amd64 GO/go1.14
+func NewRequestUserAgent() *RequestUserAgent {
+	uab := httpbinding.NewUserAgentBuilder()
+	uab.AddKeyValue(aws.SDKName, aws.SDKVersion)
+	uab.AddKeyValue("GOOS", runtime.GOOS)
+	uab.AddKeyValue("GOARCH", runtime.GOARCH)
+	uab.AddKeyValue("GO", runtime.Version())
+	if ev := os.Getenv(execEnvVar); len(ev) > 0 {
+		uab.AddKeyValue(execEnvUAKey, ev)
+	}
+	return &RequestUserAgent{uab: uab}
+}
+
+// AddUserAgentKey retrieves a RequestUserAgent from the provided stack, or initializes one.
+func AddUserAgentKey(key string) func(*middleware.Stack) error {
+	return func(stack *middleware.Stack) error {
+		requestUserAgent, err := getOrAddRequestUserAgent(stack)
+		if err != nil {
+			return err
+		}
+		requestUserAgent.AddKey(key)
+		return nil
+	}
+}
+
+// AddUserAgentKeyValue retrieves a RequestUserAgent from the provided stack, or initializes one.
+func AddUserAgentKeyValue(key, value string) func(*middleware.Stack) error {
+	return func(stack *middleware.Stack) error {
+		requestUserAgent, err := getOrAddRequestUserAgent(stack)
+		if err != nil {
+			return err
+		}
+		requestUserAgent.AddKeyValue(key, value)
+		return nil
+	}
+}
+
+func getOrAddRequestUserAgent(stack *middleware.Stack) (*RequestUserAgent, error) {
+	id := (&RequestUserAgent{}).ID()
```

Comment:
If you want you could change this to work with a nil `RequestUserAgent` value since `ID` method doesn't use receiver value.  Or alternatively, hoist the ID value to a `const` that is referenced.  Could even keep the ID and `RequestUserAgent` middleware  all as unexported to prevent direct usage outside of the exposed helpers.

```go
id := (*RequestUserAgent)(nil).ID()
```